# (PART) EXPLORATORY ANALYSIS {-}

# Review Preprocessed Data {#explore-data}

## Load libraries and data
```{r}
source("R/common.R")
library(phyloseq)
library(metagMisc)
library(microbiome)
library(microViz)

otutable_long <- readRDS("../iMAP-part3/RDataRDS/phylo_table_long.rds")
load("../iMAP-part3/RDataRDS/phyloseq_objects.RData", verbose = T)
```

Let's use objects created from `mothur` output.


## How many reads per sample
```{r}
seqcount_per_sample <- otutable_long %>% 
  group_by(sample_id) %>% 
  summarise(nseqs = sum(count), .groups = "drop") %>% 
  arrange(nseqs)

seqcount_per_sample

min_count <- min(seqcount_per_sample$nseqs)
max_count <- max(seqcount_per_sample$nseqs)

saveRDS(seqcount_per_sample, "RDataRDS/seqcount_per_sample.rds")
save(seqcount_per_sample, min_count, max_count, file = "RDataRDS/count_min_max_seqs.RData")
```


```{r}
load("../iMAP-part3/RDataRDS/phyloseq_objects.RData", verbose = T)
```


```{r}
(ps <- ps_raw)
```


## Number of taxa
```{r}
ntaxa(ps)
```

## Splitting taxon
```{r}
phyloseq_sep_tax(ps, TaxRank = "kingdom", drop_NA = FALSE)

```


## Number of samples
```{r}
nsamples(ps)
```

## Sample names
```{r}
sample_names(ps)
```

## Rank names
```{r}
rank_names(ps)
```

## Sample variables
```{r}
sample_variables(ps)
```

## Abundance table
```{r}
otu_table(ps)[1:10, 1:5]
```

## Taxonomy table
```{r}
tax_table(ps)[1:5, 1:3]
```

Phylogenetic tree
```{r}
phy_tree(ps)
```

## Taxa names
```{r}
taxa_names(ps)[1:5]
```

## Characteristics of this data
- Using the `microbiome` R package.
```{r message=FALSE, warning=FALSE}
library(microbiome)
microbiome::summarize_phyloseq(ps)

```

> Is the data composiosional? Find out fron the first line of the output from tbe previous command.

## Sort samples by abundance

> Sorting samples helps in sub-sampling.

```{r}
sort(phyloseq::sample_sums(ps), decreasing = F)
```

## Drop samples below a threshold
```{r}
ps2000 <- phyloseq::subset_samples(ps, phyloseq::sample_sums(ps) > 2000)
ps2000
```

## Drop taxa below a threshold
```{r}
pstaxa10 <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 10, ps)
pstaxa10
```

> Check if by dropping data below a specified threshold retained all sample. Adjust the threshold for desired output.

## Explore Sample Data Attributes
```{r}
ps %>% sample_data %>% dplyr::count(isolate)

```

