---
title: "Exploratory Analysis of Microbiome Data in R"
subtitle: "End-to-End Practical User Guides for Microbiome Data Analysis"
author: "Teresia Mrema-Buza, A Microbiome Data Science Enthusiast and Owner of the Complex Data Insights, LLC, USA"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
knit: bookdown::render_book
output:
  bookdown::bs4_book:
    includes:
      in_header:
    template: bs4_part3_template.html
documentclass: book
classoption: openany #remove empty pages in pdf doc
colorlinks: true
css:
- style.css
- css/style.css
always_allow_html: true
fig_caption: true
fontsize: 12pt
geometry: margin=1in
indent: false
keep_tex: true
link-citations: true
mainfont: Times New Roman
biblio-style: apalike
description: |
  This is a practical user's guide for **Systematic Microbiome Data Analysis in R**. The guide provides integrated and highly curated solutions for achieving better results.
---


```{r pkgbiblib, include=FALSE}
knitr::write_bib(c(
  .packages(), 'base','bookdown','rmarkdown','tidyverse','shiny','vegan','data.table, dendextend, robCompositions, microbiome, ALDEx2, caret, rms, phyloseq'
), 'library/packages.bib')
```

```{r setup, echo=FALSE}
source(file = "R/common.R")
if (!dir.exists("data")) {dir.create("data")}
```

# Exploratory Analysis of Microbiome Data {-#frontpage}

```{r include=FALSE}
library(DiagrammeR)
library(DiagrammeRsvg)

graph <- DiagrammeR::grViz("
digraph {
graph [layout = dot, rankdir = TD]

node [
shape = box, 
style = filled, 
fillcolor = white, 
fontname = Helvetica,
penwidth = 2.0] 

edge [arrowhead = diamond]

A [label = 'EXPLORATORY \nANALYSIS OF \nMICROBIOME DATA', fillcolor = white, penwidth = 5.0]
B [label = 'Preprocesed\nTidy Data']
C [label = 'Metadata']
D [label = 'OTU Table']
E [label = 'Taxonomy Table']
F [label = 'Composite Object']
G [label = 'Data\nTansformation']
H [label = 'Preliminary\nVisualization']
I [label = 'Deeper \nInsights...', fillcolor = yellow, penwidth = 2.0]

{A}  -> B [label = '  Input Data']
{B}  -> C
{B}  -> D
{B}  -> E
{C}  -> F
{D}  -> F
{E}  -> F
{F}  -> G
{F}  -> H
{G}  -> H
{G}  -> I


}", height = 500, width = 500)

# 2. Convert to SVG, then save as png
part3_flow = DiagrammeRsvg::export_svg(graph)
part3_flow = charToRaw(part3_flow) # flatten
rsvg::rsvg_png(part3_flow, "img/part3_flow.png")
```


<a href=""><img src="images/abund_bar.gif" alt="Book cover" width="100%" style="padding: 50px 0px 50px 0px; float: right;"/></a>

## Quick Glimpse {-}
Investigating the role of microbial communities in health and disease requires a thorough knowledge of the entire analytical process. Using wrong approaches can cost a significant amount of dollars and lengthy process to achieve the desired results. This is <b>PART 3</b> of the practical user guides intended to provide analytical support to the microbiome research community. The entire guide is reproducible, allowing users to easily follow along. If interested, user may use this model to publish their findings in a book format.

## Structure of this guide {-}
This guide is divided into chapters to facilitate easy navigation. Each chapter contains several sections as displayed in the navigation bars on the left and right. Click the hyper-linked text if you want to jump into a specific chapter or section.

## Code availability {-}
The code that generated this guide is available at a public [GitHub repository](https://github.com/tmbuza/iMAP-part3/) or via a [GH-Page](https://tmbuza.github.io/iMAP-part3/). If interested you can request a consulting service by contacting the developer of this repo using <a href="https://complexdatainsights.com/contact-us">this contact form</a>. 

<!--chapter:end:index.Rmd-->

# (PART) EXPLORATORY ANALYSIS {-}

# Exploratory Analysis of Preprocessed Data {#explore-data}

## Prerequisite
- Data objects containing metadata, abundance and taxonomy such as **R and phyloseq objects** prepared in previous section.
- Analysis and visualization tools

## Load libraries and data
```{r}
source("R/common.R")
library(phyloseq)
library(metagMisc)
library(microbiome)
library(microViz)

load("../iMAP_part3/RDataRDS/phyloseq_objects.RData", verbose = T)
```

Let's use phyloseq object created from `mothur` output. 

```{r}
(ps <- ps_mo_raw)
```


## Number of taxa
```{r}
ntaxa(ps)
```

## Splitting taxon
```{r}
phyloseq_sep_tax(ps, TaxRank = "kingdom", drop_NA = FALSE)

```


## Number of samples
```{r}
nsamples(ps)
```

## Sample names
```{r}
sample_names(ps)
```

## Rank names
```{r}
rank_names(ps)
```

## Sample variables
```{r}
sample_variables(ps)
```

## Abundance table
```{r}
otu_table(ps)[1:10, 1:5]
```

## Taxonomy table
```{r}
tax_table(ps)[1:5, 1:3]
```

Phylogenetic tree
```{r}
phy_tree(ps)
```

## Taxa names
```{r}
taxa_names(ps)[1:5]
```

## Characteristics of this data
- Using the `microbiome` R package.
```{r message=FALSE, warning=FALSE}
library(microbiome)
microbiome::summarize_phyloseq(ps)

```

> Is the data composiosional? Find out fron the first line of the output from tbe previous command.

## Sort samples by abundance

> Sorting samples helps in sub-sampling.

```{r}
sort(phyloseq::sample_sums(ps), decreasing = F)
```

## Drop samples below a threshold
```{r}
ps2000 <- phyloseq::subset_samples(ps, phyloseq::sample_sums(ps) > 2000)
ps2000
```

## Drop taxa below a threshold
```{r}
pstaxa10 <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 10, ps)
pstaxa10
```

> Check if by dropping data below a specified threshold retained all sample. Adjust the threshold for desired output.

## Explore Sample Data Attributes
```{r}
ps %>% sample_data %>% dplyr::count(isolate)

```


<!--chapter:end:04a_data_exploration.Rmd-->

# (APPENDIX) APPENDIX {-}

# Saved Data Objects

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```


## RDS Format for Individual Object
```{r}

rds <- list.files(path="./RDataRDS", pattern = "*.rds", full.names = TRUE)
rds

```

## RData Format for Multiple Objects
```{r}

rdata <- list.files(path="./RDataRDS", pattern = "*.RData", full.names = TRUE)
rdata

```


## CSV or TSV Format Files
```{r}

list.files(path="./RDataRDS", pattern = "*.csv|tsv", full.names = TRUE)

```


```{r eval=FALSE, include=FALSE}
## All-in-One Input-Output Data
lsdata("RDataRDS/saved_objects.RData")
```


## How to reload RDS or RData
```{block, type="tmbinfo", echo=T}
### RDS format e.g. foo.rds
foo <-`readRDS("RDataRDS/foo.rds")`

### RData format e.g. foo.RData
`load("RDataRDS/foo.RData", verbose = TRUE)`

### List objects in RData
`lsdata("foo.RData")`
```



# Software and Packages

## Basic dependencies
* `r R.version.string`
* `tidyverse` (v. `r packageVersion("tidyverse")`)
* `knitr` (v. `r packageVersion("knitr")`)
* `rmarkdown` (v. `r packageVersion("rmarkdown")`)
* `bookdown` (v. `r packageVersion("bookdown")`)

## Available on machine used 
```{r}
sessionInfo()

```




<!--chapter:end:99_appendix.Rmd-->

`r if (knitr:::is_html_output()) '
# References {-}
'`


<!--chapter:end:999-references.Rmd-->

